ICE: trait takes 1 generic argument but 0 generic arguments were supplied

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

I tried this code.

### Code

```Rust
pub trait Super<X> {
    type X;
}

impl<T> Clone for Box<dyn Super<X = T>> {
    fn clone(&self) -> Self {
        unimplemented!();
    }
}

pub fn main() {}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.92.0-nightly (4068bafed 2025-10-20)
binary: rustc
commit-hash: 4068bafedd8ba724e332a5221c06a6fa531a30d2
commit-date: 2025-10-20
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.3
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program Output</strong></summary>
<p>

```
error: internal compiler error[E0107]: trait takes 1 generic argument but 0 generic arguments were supplied
 --> <source>:5:27
  |
5 | impl<T> Clone for Box<dyn Super<X = T>> {
  |                           ^^^^^ expected 1 generic argument
  |
note: trait defined here, with 1 generic parameter: `X`
 --> <source>:1:11
  |
1 | pub trait Super<X> {
  |           ^^^^^ -
note: delayed at compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs:1146:23 - disabled backtrace
 --> <source>:5:27
  |
5 | impl<T> Clone for Box<dyn Super<X = T>> {
  |                           ^^^^^
help: add missing generic argument
  |
5 | impl<T> Clone for Box<dyn Super<X, X = T>> {
  |                                 ++

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:649:29 - disabled backtrace

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/app/rustc-ice-2025-10-26T02_27_23-1.txt` to your bug report

note: compiler flags: -C debuginfo=2 --crate-type bin -C linker=/opt/compiler-explorer/gcc-15.2.0/bin/gcc

query stack during panic:
end of query stack
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/generics.rs
@@ -9,6 +9,7 @@ use rustc_middle::ty::{
 };
 use rustc_session::lint::builtin::LATE_BOUND_LIFETIME_ARGUMENTS;
 use rustc_span::kw;
+use rustc_trait_selection::traits;
 use smallvec::SmallVec;
 use tracing::{debug, instrument};
 
@@ -535,9 +536,26 @@ pub(crate) fn check_generic_arg_count(
                     .map(|param| param.name)
                     .collect();
                 if constraint_names == param_names {
+                    let has_assoc_ty_with_same_name =
+                        if let DefKind::Trait = cx.tcx().def_kind(def_id) {
+                            gen_args.constraints.iter().any(|constraint| {
+                                traits::supertrait_def_ids(cx.tcx(), def_id).any(|trait_did| {
+                                    cx.probe_trait_that_defines_assoc_item(
+                                        trait_did,
+                                        ty::AssocTag::Type,
+                                        constraint.ident,
+                                    )
+                                })
+                            })
+                        } else {
+                            false
+                        };
                     // We set this to true and delay emitting `WrongNumberOfGenericArgs`
-                    // to provide a succinct error for cases like issue #113073
-                    all_params_are_binded = true;
+                    // to provide a succinct error for cases like issue #113073,
+                    // but only if when we don't have any assoc type with the same name with a
+                    // generic arg. Otherwise it will cause an ICE due to a delayed error because we
+                    // don't have any error other than `WrongNumberOfGenericArgs`.
+                    all_params_are_binded = !has_assoc_ty_with_same_name;
                 };
             }
 

--- a/tests/ui/traits/associated_type_bound/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.rs
+++ b/tests/ui/traits/associated_type_bound/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.rs
@@ -1,0 +1,17 @@
+// A regression test for https://github.com/rust-lang/rust/issues/148121
+
+pub trait Super<X> {
+    type X;
+}
+
+pub trait Zelf<X>: Super<X> {}
+
+pub trait A {}
+
+impl A for dyn Super<X = ()> {}
+//~^ ERROR: trait takes 1 generic argument but 0 generic arguments were supplied
+
+impl A for dyn Zelf<X = ()> {}
+//~^ ERROR: trait takes 1 generic argument but 0 generic arguments were supplied
+
+fn main() {}

--- a/tests/ui/traits/associated_type_bound/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.stderr
+++ b/tests/ui/traits/associated_type_bound/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.stderr
@@ -1,0 +1,35 @@
+error[E0107]: trait takes 1 generic argument but 0 generic arguments were supplied
+  --> $DIR/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.rs:11:16
+   |
+LL | impl A for dyn Super<X = ()> {}
+   |                ^^^^^ expected 1 generic argument
+   |
+note: trait defined here, with 1 generic parameter: `X`
+  --> $DIR/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.rs:3:11
+   |
+LL | pub trait Super<X> {
+   |           ^^^^^ -
+help: add missing generic argument
+   |
+LL | impl A for dyn Super<X, X = ()> {}
+   |                      ++
+
+error[E0107]: trait takes 1 generic argument but 0 generic arguments were supplied
+  --> $DIR/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.rs:14:16
+   |
+LL | impl A for dyn Zelf<X = ()> {}
+   |                ^^^^ expected 1 generic argument
+   |
+note: trait defined here, with 1 generic parameter: `X`
+  --> $DIR/assoc-type-bounds-with-the-same-name-with-lacking-generic-arg-148121.rs:7:11
+   |
+LL | pub trait Zelf<X>: Super<X> {}
+   |           ^^^^ -
+help: add missing generic argument
+   |
+LL | impl A for dyn Zelf<X, X = ()> {}
+   |                     ++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0107`.

